packages/core/src/util/txInspector.ts

```diff
@@ -0,0 +1,145 @@
+  StakeDelegationCertificate,
+  TokenMap,
+  TxAlonzo,
+  TxOut,
+  Value
+} from '../Cardano';
+
+const isMyAddress =
+  (addresses: Address[]) =>
+  ({ address }: { address: Address }) =>
+    addresses.includes(address!);
+
+const isOutgoing = (tx: TxAlonzo, ownAddresses: Address[]): boolean => tx.body.inputs.some(isMyAddress(ownAddresses));
+
+const getTotalValue = (outputs: TxOut[]): Value => {
+  const result: Value = { coins: 0n };
+  const totalAssets: TokenMap = new Map();
+
+  for (const output of outputs) {
+    result.coins += output.value.coins;
+    if (output.value.assets) {
+      for (const [asset, amount] of output.value.assets) {
+        const total = totalAssets.get(asset) ?? 0n;
+        totalAssets.set(asset, total + amount);
+      }
+    }
+  }
+
+  if (totalAssets.size > 0) result.assets = totalAssets;
+  return result;
+};
+
+type InspectionResult = { [k: string]: unknown } | null;
+type InspectTx = (tx: TxAlonzo) => InspectionResult;
+type Inspectors = { [k: string]: InspectTx };
+type TxInspector<T extends Inspectors> = (tx: TxAlonzo) => {
+  [k in keyof T]: ReturnType<T[k]>;
+};
+
+export type SendReceiveCoinInspector = {
+  coins: Lovelace;
+  assets?: TokenMap;
+} | null;
+
+export type DelegateInspector = {
+  stakePool: PoolId;
+  stakeKeyHash: Ed25519KeyHash;
+} | null;
+
+export type StakeKeyRegistrationInspector = {
+  stakeKeyHash: Ed25519KeyHash;
+} | null;
+
+/**
+ * Inspects a transaction for coins and assets sent from the provided addresses.
+ * Returns `null` if it's not sending anything.
+ *
+ * @param {Address[]} ownAddresses own wallet's addresses
+ */
+export const sendCoinInspector =
+  (ownAddresses: Address[]) =>
+  (tx: TxAlonzo): SendReceiveCoinInspector => {
+    if (!isOutgoing(tx, ownAddresses)) return null;
+    const sentOutputs = tx.body.outputs.filter((out) => !isMyAddress(ownAddresses)(out));
+    return getTotalValue(sentOutputs);
+  };
+
+/**
+ * Inspects a transaction for coins and assets received by the provided addresses.
+ * Returns `null` if it's not receiving anything.
+ *
+ * @param {Address[]} ownAddresses own wallet's addresses
+ */
+export const receiveCoinInspector =
+  (ownAddresses: Address[]) =>
+  (tx: TxAlonzo): SendReceiveCoinInspector => {
+    if (isOutgoing(tx, ownAddresses)) return null;
+    const receivedOutputs = tx.body.outputs.filter((out) => isMyAddress(ownAddresses)(out));
+    return getTotalValue(receivedOutputs);
+  };
+
+/**
+ * Inspects a transaction for a stake delegation certificate.
+ * Returns `null` if the transaction doesn't have one.
+ *
+ * @param {TxAlonzo} tx transaction to inspect
+ */
+export const delegationInspector = (tx: TxAlonzo): DelegateInspector => {
+  const delegation = tx.body.certificates?.find((cert) => cert.__typename === CertificateType.StakeDelegation) as
+    | StakeDelegationCertificate
+    | undefined;
+  return delegation ? { stakeKeyHash: delegation.stakeKeyHash, stakePool: delegation.poolId } : null;
```

Comment:
Could a tx have more than one certificate of the same type? If so, should we return an array instead?

Reply:
Yes, return type should be an array

Reply:
@rhyslbw @mkazlauskas In case there are no certificates and considering we are removing `null` for the send/receive ones, what do you think it's best to return: `null` or just an empty array?

Reply:
I think both options are ok, just go with whichever you think would be easier or more intuitive to consume in the wallet

Reply:
Done in 19ae735f102d9a785ed2ab7a84f055380eb1ce77, removed all nulls